pkg/controller/volume/attachdetach/cache/actual_state_of_world.go

```diff
@@ -324,17 +327,20 @@ func (asw *actualStateOfWorld) SetVolumeMountedByNode(
 	if err != nil {
 		return fmt.Errorf("Failed to SetVolumeMountedByNode with error: %v", err)
 	}
+	// TODO: probably safer to switch getNodeAndVolume to return pointers?
 
 	if mounted {
 		// Increment set count
 		nodeObj.mountedByNodeSetCount = nodeObj.mountedByNodeSetCount + 1
 	} else {
 		// Do not allow value to be reset unless it has been set at least once
+		// TODO: not sure I understand the logic here?
 		if nodeObj.mountedByNodeSetCount == 0 {
 			return nil
 		}
 	}
 
+	// TODO-P0: volumeObj could be attachedVolume{} (if volume not found)
```

Comment:
I think it is possible, although very unlikely. Right before ResetDetachRequestTime is called, reconciler checks the volume is in both desired state and actual state. Only if right before this, desired state does not have the volume, and reconciler sends an asynchronous detach operation and it finishes right before this ResetDetachRequestTime() call and after checking desired and actual state, any empty list attachedVolume{} might be returned. 
But you are right, we should check before updating the value.